src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,674 @@
+        return 0;
+    }
+
+    /* Serialize B_m */
+    secp256k1_ge_set_gej(&result_ge, &result_gej);
+    secp256k1_pubkey_save(labelled_spend_pubkey, &result_ge);
+
+    return 1;
+}
+
+int secp256k1_silentpayments_recipient_public_data_create(
+    const secp256k1_context *ctx,
+    secp256k1_silentpayments_recipient_public_data *public_data,
+    const unsigned char *outpoint_smallest36,
+    const secp256k1_xonly_pubkey * const *xonly_pubkeys,
+    size_t n_xonly_pubkeys,
+    const secp256k1_pubkey * const *plain_pubkeys,
+    size_t n_plain_pubkeys
+) {
+    size_t i;
+    size_t pubkeylen = 65;
+    secp256k1_ge A_sum_ge, addend;
+    secp256k1_gej A_sum_gej;
+    unsigned char input_hash_local[32];
+    int ret = 1;
+    int ser_ret;
+
+    /* Sanity check inputs */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(outpoint_smallest36 != NULL);
+    ARG_CHECK((plain_pubkeys != NULL) || (xonly_pubkeys != NULL));
+    if (xonly_pubkeys != NULL) {
+        ARG_CHECK(n_xonly_pubkeys > 0);
+    } else {
+        ARG_CHECK(n_xonly_pubkeys == 0);
+    }
+    if (plain_pubkeys != NULL) {
+        ARG_CHECK(n_plain_pubkeys > 0);
+    } else {
+        ARG_CHECK(n_plain_pubkeys == 0);
+    }
+    secp256k1_memclear(input_hash_local, 32);
+
+    /* Compute input public keys sum: A_sum = A_1 + A_2 + ... + A_n */
+    secp256k1_gej_set_infinity(&A_sum_gej);
+    for (i = 0; i < n_plain_pubkeys; i++) {
+        ret &= secp256k1_pubkey_load(ctx, &addend, plain_pubkeys[i]);
+        secp256k1_gej_add_ge_var(&A_sum_gej, &A_sum_gej, &addend, NULL);
+    }
+    for (i = 0; i < n_xonly_pubkeys; i++) {
+        ret &= secp256k1_xonly_pubkey_load(ctx, &addend, xonly_pubkeys[i]);
+        secp256k1_gej_add_ge_var(&A_sum_gej, &A_sum_gej, &addend, NULL);
+    }
+    /* Since an attacker can maliciously craft transactions where the public keys sum to zero, fail early here
+     * to avoid making the caller do extra work, e.g., when building an index or scanning many malicious transactions
+     */
+    if (secp256k1_gej_is_infinity(&A_sum_gej)) {
+        return 0;
+    }
+    /* Compute input_hash = hash(outpoint_L || A_sum) */
+    secp256k1_ge_set_gej(&A_sum_ge, &A_sum_gej);
+    secp256k1_silentpayments_calculate_input_hash(input_hash_local, outpoint_smallest36, &A_sum_ge);
+    /* serialize the public_data struct */
+    public_data->data[0] = 0;
+    ser_ret = secp256k1_eckey_pubkey_serialize(&A_sum_ge, &public_data->data[1], &pubkeylen, 0);
+    VERIFY_CHECK(ser_ret && pubkeylen == 65);
+    (void)ser_ret;
+    memcpy(&public_data->data[1 + pubkeylen], input_hash_local, 32);
+    return ret;
+}
+
+static int secp256k1_silentpayments_recipient_public_data_load_pubkey(const secp256k1_context *ctx, secp256k1_pubkey *pubkey, const secp256k1_silentpayments_recipient_public_data *public_data) {
+    size_t pubkeylen = 65;
+    return secp256k1_ec_pubkey_parse(ctx, pubkey, &public_data->data[1], pubkeylen);
```

Comment:
`ec_pubkey_parse` is for parsing arbitrary byte arrays and includes a relatively expensive check that the coordinates are valid. It would be better to use `ge_from_bytes` and `ge_to_bytes` similar to the `*_load` functions in the musig module. This would also reduce the size of the the public data by a byte.